### PR TITLE
Add chunking support to linear_cross_entropy to control memory usage

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

## Benchmarks

The figure below contains benchmarks (memory usages and timings) with a number of linear cross entropy implementations listed below:
- `LinearCrossEntropyLoss[options=None]` corresponds to the naive implementation (green lines),
- `LiLinearCrossEntropyLoss` corresponds to [Liger-Kernel](https://github.com/linkedin/Liger-Kernel) implementation (orange lines),
- `FusedProjectionPlusCrossEntropyLoss[n_loop_iters=8]` corresponds to [efficient_cross_entropy](https://github.com/mgmalek/efficient_cross_entropy) implementation (blue lines),
- `VoLinearCrossEntropyLoss` corresponds to vocab implementation (purple lines),
- `NewLinearCrossEntropyLoss[max_memory_gb=1.5 grad_inplace=True]`  corresponds to the implementation of this PR (red lines) using chunking under the constraint that memory usage should not exceed 1.5 GB (this corresponds to the horizontal part of the curve). If the problem does not fit to this limit, the memory usage will be close to the theoretical minima (black dashed-dotted line) that is defined by the storage size of inputs and gradients.

NVIDIA_GeForce_RTX_2060_SUPER-CC7 5-7786MB-sum-indices

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* __->__ #175039
* #172446
* #172286

### CODE REVIEW

File: torch/nn/modules/_functions.py [hunk 1]
```diff
@@ -317,3 +319,570 @@ def forward(ctx, *args):
...
+
+        min_chunk_size: int = opt_options.get("min_chunk_size", 1024)
+        max_memory_gb = opt_options.get("max_memory_gb")
+        if max_memory_gb is None:
+            if device.type == "cuda":
+                max_memory_gb = int(torch.cuda.mem_get_info(device)[0] * 0.75 / 1e9)
+            else:
+                max_memory_gb = 64
+
+        max_total_numel = int(max_memory_gb * 1e9 / dtype.itemsize)
```
Comment (Skylion007) [inline torch/nn/modules/_functions.py]: Where is this mat_total_num 1e9 constant derived from?

File: torch/nn/modules/_functions.py [hunk 2]
```diff
@@ -317,3 +319,570 @@ def forward(ctx, *args):
...
+                return opt_options
+
+        constraints = [f"{num_batches=}, {in_features=}, {num_classes=}"]
+        for k, v in opt_options.items():
+            if k not in {"max_memory_gb", "min_chunk_size"}:
+                constraints.append(f"{k}={v}")
+        constraints = ", ".join(constraints)
+        warnings.warn(
+            "failed to find optimal chunking strategy for linear_cross_entropy: "
+            f"{max_memory_gb=} is too small or {min_chunk_size=} is too large. Constraints: {constraints}"
```
Comment (Skylion007) [inline torch/nn/modules/_functions.py]: Please add a stacklevel arg here...

File: torch/nn/functional.py [hunk 3]
```diff
@@ -3692,7 +3694,14 @@ def linear_cross_entropy(
             Architecture for Computer Vision
             <https://arxiv.org/abs/1512.00567>`__.
             Default: :math:`0.0`.
-
+        options (dict, optional): Specify chunking strategy options,
```
Comment (Skylion007) [inline torch/nn/functional.py]: Why an options duct in the first place instead of something more structural like a dataclass or TypedDict?

File: torch/nn/functional.py [hunk 4]
```diff
@@ -3728,18 +3738,71 @@ def linear_cross_entropy(
...
             reduction=reduction,
             label_smoothing=label_smoothing,
+            options=options,
+        )
+    out_features = linear_weight.shape[:-2]
+    num_classes = linear_weight.shape[-2]
+    in_features = input.shape[-1]
+    if out_features:
+        linear_weight = linear_weight.reshape(
+            (num_classes * math.prod(out_features), in_features)
```
Comment (Skylion007) [inline torch/nn/functional.py]: Nit: you can optimize this as math.prod(num_features, start=num_classes)

File: torch/nn/modules/_functions.py (hunk 1 above)
Comment (pearu) [reply torch/nn/modules/_functions.py]: 1e9 is an approximation of 2 ** 30 to convert giga-bytes to bytes, it does not have to be exact as `max_memory_gb` itself is an approximate parameter.

File: torch/nn/modules/_functions.py (hunk 2 above)
Comment (pearu) [reply torch/nn/modules/_functions.py]: Done.

File: torch/nn/functional.py (hunk 4 above)
Comment (pearu) [reply torch/nn/functional.py]: Applied. Thanks!

File: torch/nn/functional.py (hunk 3 above)
Comment (pearu) [reply torch/nn/functional.py]: Added `LinearCrossEntropyOptions(TypedDict)`.

File: torch/nn/modules/_functions.py (hunk 1 above)
Comment (pearu) [reply torch/nn/modules/_functions.py]: Added a constant `gb_to_bytes = 2 ** 30` for unit conversions.

**Review by mikaylagawarecki — COMMENTED** [review]
I want to preface this by saying that I deeply appreciate the effort that you have put into this from deriving the backward from first principles, to reproducing the benchmark results of efficient cross entropy to the thought process required to getting this PR out.

Based on discussion, we would like this PR to be broken down into a stack of 3 PRs in this order
1.  A PR that includes the batch chunking logic
2. A PR that includes the vocab chunking logic (mutually exclusive with batch chunking) that matches the recompute decision of the provided reference
3. [For discussion] This PR that includes the `max_memory_gb` logic + chunking along all 3 dims

 It is an explicit goal on our side to provide one implementation that chunks along batch dimension and one implementation that chunks along the vocab dimension (with recompute in backward) in core. I think you are mostly aligned with us on the value of chunking along the batch dim, but I understand you have qualms about the vocab chunking implementation. We do consider the vocab chunking implementation to be battle tested and I attempt to address your concerns in the extra exposition section. 


We consider the implementation that derives optimal chunking by estimating memory and using heuristics to chunk along all 3 dimensions simultaneously as an optimization. From our preliminary discussion, we think this requires more alignment. For example, there were differing opinions re `max_memory_gb` (while there is merit since it is the last op in forward, the memory estimation lacks precedence in eager and seems closer to what is usually done with graph passes in the compiler)


Concretely this is what we are looking for in the first two PRs

1.  PR that chunks along batch_dim (focus on landing this one first)
- `LinearCrossEntropyOptions` is a dataclass that only has `grad_inplace` and `batch_chunk_size`.
- Profiler trace that shows Liger in the PR description, profiler trace that shows your impl in the PR description
- Benchmarks  show memory usage and performance on par with liger kernel with benchmark script in PR description (iiuc memory usage is already on par and from the perf benchmarks you shared earlier your batch chunking implementation already had performance on par, though (see my question below) I don't understand how this implementation has perf on par)
- A screenshot of the memory snapshot in the PR description comparing naive linear + cross_entropy to batch chunking see https://docs.pytorch.org/docs/stable/torch_cuda_memory.html for how to use memory snapshot
- OpInfo tests to exercise this path (iiuc they currently don't)
- re: composite compliance, iiuc, `autograd.Function` in python cannot be composite compliant as the appropriate levers do not exist to make it composite compliant, there are two options here
(a) Use a python `custom_op` instead
(b) Register a private function to native_functions.yaml/derivatives.yaml 
If you can do everything in python, I vote for the first. But if we do need to replicate the fusions that liger does (see my q below) then you might need to go with the second.

2. PR that chunks along vocab_dim
- `class_chunk_size` is added to `LinearClassEntropyOptions`, this is mutually exclusive with `batch_chunk_size` (i.e. we cannot chunk on both dims at the same time)
- CUDA kernels are written to fuse the parts of the kernel that are `torch.compile`d (the fwd and bwd epilogue)
- Benchmarks  show memory usage and performance on par with vocab chunking implementation 
- Ditto for memory snapshot/OpInfo
- You probably have to register a private `_vocab_chunking_impl` to native_functions.yaml here since this requires CUDA code


### More Exposition

This op is mostly intended to be used in the context of LLMs. Here are some ballpark numbers I got from Claude on what we can expect the sizes to look like  
```
  ┌────────────────┬────────────┬──────────────┬──────────────┐                                                                                                                   
  │                │ Small (7B) │ Medium (70B) │ Large (405B) │                                                                                                                   
  ├────────────────┼────────────┼──────────────┼──────────────┤                                                                                                                   
  │ Hidden dim (H) │ 4096       │ 8192         │ 16384        │                                                                                                                   
  ├────────────────┼────────────┼──────────────┼──────────────┤
  │ Vocab size (V) │ 32K–128K   │ 32K–128K     │ 128K         │
  ├────────────────┼────────────┼──────────────┼──────────────┤
  │ Seq length (T) │ 2K–8K      │ 4K–128K      │ 8K–128K      │
  └────────────────┴────────────┴──────────────┴──────────────┘
```

  For training, num_tokens (BT) = batch_size × seq_length. Typical values:

  - Micro-batch: 1–4 sequences per GPU
  - Seq length: 4096–8192 is common, up to 128K for long-context
  - So BT per GPU: roughly 4K–32K tokens in a micro-batch, sometimes up to 128K+ 


To my knowledge, I don’t think there is an implementation in open source that chunks along the in_features (hidden_dim) dimension. I suspect this is because in the context of LLMs the hidden_dim is usually much smaller than token/vocab and there are limited benefits to reap from chunking further without reducing GEMM efficiency. To reduce complexity, let us remove chunking along this dim for now and we can add it in future if it is requested.


Re vocab chunking, responding to your message from the chat

> Even it works, I am not a fan of it: it requires extra operations to compute the final result and numerically it is likely less accurate due to repeated exp/log operations when adjusting the logits against updated max values; and the algorithm is more complex. I think an equivalent alternative to reducing memory usage is to use more chunking along other dimensions that has much less processing overhead compared to vocab chunking approach.

- numerically it is likely less accurate due to repeated exp/log operations when adjusting the logits against updated max values —> the authors say that the numerics of the standard PyTorch reference are preserved exactly
- requires extra operations to compute the final result —> this is an explicit tradeoff that the authors choose to make
- given the ballparks for vocab/batch size above, vocab dim allows us to reduce peak memory more as it can be chunked more


### Questions
1. To be honest, something I'm not understanding is how you were able to get performance on par for batch chunking in the previous screenshot you sent me without needing a fused kernel like what is done [here](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/cross_entropy.py#L27), could you help me understand that?
image


You should benchmark with larger vocab size, if your GPU cannot handle that can we check with Andrew James how you can get one?
 
 2. https://github.com/openteams-ai/pearu-sandbox/blob/master/pytorch/linear_cross_entropy/benchmark.py#L253 It looks like maybe the benchmark is not doing warmup? Maybe we can use [`triton.testing.do_bench` ](https://triton-lang.org/main/python-api/generated/triton.testing.do_bench.html) as the efficient_cross_entropy repo does
